retrieval/discovery/azure.go

```diff
@@ -0,0 +1,214 @@
+		ch <- []*config.TargetGroup{tg}
+	}
+
+	for {
+		select {
+		case <-ticker.C:
+			tg, err := ad.refresh()
+			if err != nil {
+				log.Error(err)
+			} else {
+				ch <- []*config.TargetGroup{tg}
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// azureClient represents multiple Azure Resource Manager providers.
+type azureClient struct {
+	nic network.InterfacesClient
+	vm  compute.VirtualMachinesClient
+}
+
+// createAzureClient is a helper function for creating an Azure compute client to ARM.
+func createAzureClient(cfg config.AzureSDConfig) (azureClient, error) {
+	var c azureClient
+	oauthConfig, err := azure.PublicCloud.OAuthConfigForTenant(cfg.TenantID)
+	if err != nil {
+		return azureClient{}, err
+	}
+	spt, err := azure.NewServicePrincipalToken(*oauthConfig, cfg.ClientID, cfg.ClientSecret, azure.PublicCloud.ResourceManagerEndpoint)
+	if err != nil {
+		return azureClient{}, err
+	}
+
+	c.vm = compute.NewVirtualMachinesClient(cfg.SubscriptionID)
+	c.vm.Authorizer = spt
+
+	c.nic = network.NewInterfacesClient(cfg.SubscriptionID)
+	c.nic.Authorizer = spt
+
+	return c, nil
+}
+
+func (ad *AzureDiscovery) refresh() (*config.TargetGroup, error) {
+	tg := &config.TargetGroup{}
+	client, err := createAzureClient(*ad.cfg)
+	if err != nil {
+		return tg, fmt.Errorf("could not create Azure client: %s", err)
+	}
+
+	var machines []compute.VirtualMachine
+	result, err := client.vm.ListAll()
+	if err != nil {
+		return tg, fmt.Errorf("could not list virtual machines: %s", err)
+	}
+	machines = append(machines, *result.Value...)
+
+	// If we still have results, keep going until we have no more.
+	for result.NextLink != nil {
+		result, err = client.vm.ListAllNextResults(result)
+		if err != nil {
+			return tg, fmt.Errorf("could not list virtual machines: %s", err)
+		}
+		machines = append(machines, *result.Value...)
+	}
+	log.Debugf("Found %d virtual machines during Azure discovery.", len(machines))
+
+	// We have the slice of machines. Now turn them into targets.
+	// Doing them in go routines because the network interface calls are slow.
+	type target struct {
+		labelSet model.LabelSet
+		err      error
+	}
+
+	ch := make(chan target, len(machines))
+	for i, vm := range machines {
+		go func(i int, vm compute.VirtualMachine) {
+			resourceGroup := strings.ToLower(strings.Split(*vm.ID, "/")[4])
+			labels := model.LabelSet{
+				azureLabelMachineID:            model.LabelValue(*vm.ID),
+				azureLabelMachineName:          model.LabelValue(*vm.Name),
+				azureLabelMachineLocation:      model.LabelValue(*vm.Location),
+				azureLabelMachineResourceGroup: model.LabelValue(resourceGroup),
+			}
+
+			if vm.Tags != nil {
+				for k, v := range *vm.Tags {
+					name := strutil.SanitizeLabelName(k)
+					labels[azureLabelMachineTag+model.LabelName(name)] = model.LabelValue(*v)
+				}
+			}
+
+			// Get the IP address information via seperate call to the network provider.
+			for _, nic := range *vm.Properties.NetworkProfile.NetworkInterfaces {
```

Comment:
I'm still new to Go but if I understand the current code correctly, a non-nil labelset (target) should only be sent on line 199 if no errors were encountered.

A nil labelset and a non-nil error is being sent when an error is found and this is checked on line 206 to ensure it's an all or nothing deal.

I think there may be an issue on line 185 which you pointed out in another comment. In that particular case, I don't want to send a target because if a machine is de-allocated, it cannot be monitored but I also don't want to send an error because this may be normal. In our case, we de-allocate large chunks of our QA environment to save money and re-allocate as needed.

Do you think for the case on line 185 I should send a nil labelset and check that at ~206 instead of the naked return?